### PR TITLE
(PUP-6056) Specify 4x function API loading

### PR DESCRIPTION
**Pull request by hlindberg**
This specifies how 4.x function loading works; what is supported in
terms of creating name spaced functions and the paths from which they
are loaded.

### CODE REVIEW

File: language/func-api.md [hunk 1]
```diff
@@ -37,6 +37,9 @@ It is not uncommon that 80% of the logic in  function consists of argument type
 is when there is no checking at all (because it is a chore to write) leading to mysterious
 and sometimes spectacular failures.
 
+For autoloading information see [Autoloading][1].
```
Comment (ripienaar) [inline language/func-api.md]: this don't render right it should be a lowercase in the link and uppercase in text, maybe try `[Autoloading](#autoloading)`?

File: language/func-api.md [hunk 2]
```diff
@@ -539,3 +544,33 @@ called directly from Ruby, then the default applies, if called from the puppet l
...
+
+For the 4.x Puppet API it is not possible to autoload top scoped functions. It is however
+possible to create top scoped functions by defining them in a manifest that is guaranteed to be loaded first (typically site.pp). This mechanism can be used to override/patch and delegate function calls as a version migration aid. 
+
+Manifest loaded functions can define functions in any namespace. This should only be used for special cases (migration / patching).
+
+Nested namespaces are suppoted in the 4.x API for both Ruby and Puppet. The paths to the `.rb` or `.pp` files containing such functions should have each additional namespace in a nested directory. As an example the function `environment::testing::env_func()` should be placed in `<root>/lib/puppet/functions/environment/testing/env_func.rb` (Ruby API), or `<root>/functions/testing/env_func.pp` (Puppet API) - note the difference that the Ruby API path includes the `environment` directory since the Ruby API allows top scope functions while the Puppet API does not.
+
+
+[1]: #Autoloading
```
Comment (ripienaar) [inline language/func-api.md]: or maybe just make `#autoloading` here and leave above as is

File: language/func-api.md [hunk 3]
```diff
@@ -539,3 +544,33 @@ called directly from Ruby, then the default applies, if called from the puppet l
...
+
+| API        | in       | namespace support     | allows top-scope | path |
+| ---        | ---      | ------------------    | ---------------- | --- |
+| 3.x Ruby   | module   | Not supported         | Yes (always top scope) | `<root>/lib/puppet/parser/functions/<function_name>.rb` |
+|            | env      | Not supported         | -   | - 
+| 4.x Ruby   | module   | Yes `<module name>::` | Yes | `<root>/lib/puppet/functions/<module name>/<function_name>.rb` |
+|            | env      | Yes `environment::`   | Yes | `<root>/lib/puppet/functions/environment/<function_name>.rb` `<root>/lib/puppet/functions/<function_name>.rb` |
+| 4.x Puppet | module   | Yes `<module name>::` | No  | `<root>/functions/<function_name>.pp` |
+|            | env      | Yes `environment::`   | No  | `<root>/functions/<function_name>.pp` |
+|            | manifest | Yes (any namspace)    | Yes | in any manifest |
```
Comment (hlindberg) [inline language/func-api.md]: Yes, you can hurt yourself with that - but not any worse that how you can hurt yourself with the rest. It is also very valuable as it allows you to create functions that override and delegate as a means to overcome version differences. A cool example was adding puppet functions for the `hiera_xxx` functions that delegates to the `lookup` function by placing those functions in the `site.pp`.

File: language/func-api.md (hunk 2 above)
Comment (hlindberg) [inline language/func-api.md]: ok, thanks

File: language/func-api.md (hunk 1 above)
Comment (hlindberg) [inline language/func-api.md]: Fixed at the bottom of the document.

**Pull request merged**